proficio_sandbox/mirror_image.cpp

```diff
@@ -0,0 +1,252 @@
+ * @file mirror_image.cpp
+ *
+ * @date 02/22/10
+ * @author Christopher Dellin
+ * @author Dan Cody
+ * @author Brian Zenowich
+ */
+/**
+ * Revision History
+ * 07/31/14 - JH - Altered to Work for Proficio and Added Exit Option.
+ */
+
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include <boost/thread.hpp>
+
+#include <barrett/os.h>
+#include <barrett/detail/stl_utils.h>
+#include <barrett/units.h>
+#include <barrett/systems.h>
+#include <barrett/products/product_manager.h>
+
+#define BARRETT_SMF_VALIDATE_ARGS
+#include <barrett/standard_main_function.h>
+
+#include "mirror_image.h"
+
+
+using namespace barrett;
+using detail::waitForEnter;
+
+/* Validate CommandLine Arguments */
+bool validate_args(int argc, char** argv) {
+	if (argc != 2  &&  argc != 3) {
+		printf("Usage: %s <remoteHost> [--auto]\n", argv[0]);
+		printf("  --auto : Automatically link WAMs and start Hand or Gimbals Hand Controller thread\n");
+
+		return false;
+	}
+
+	return true;
+}
+
+/* Main Function */
+template<size_t DOF>
+int wam_main(int argc, char** argv, ProductManager& pm, systems::Wam<DOF>& wam) {
+
+	BARRETT_UNITS_TEMPLATE_TYPEDEFS(DOF);
+
+	jp_type SYNC_POS(wam.getHomePosition());  // the position each WAM should move to before linking
+	SYNC_POS(0) = 0.0; SYNC_POS(1) = 0.0;
+
+	MasterMaster<DOF> mm(pm.getExecutionManager(), argv[1]);
+	systems::connect(wam.jpOutput, mm.input);
+
+	wam.gravityCompensate();
+
+	std::vector<std::string> autoCmds;
+	autoCmds.push_back("l");
+
+
+	std::string line, qV;
+	v_type gainTmp;
+	bool active = true;
+	while (active) {
+		if (autoCmds.empty()) {
+			printf(">>> ");
+			std::getline(std::cin, line);
+		} else {
+			line = autoCmds.back();
+			autoCmds.pop_back();
+		}
+
+		switch (line[0]) {
+		case 'l':
+			if (mm.isLinked()) {
+				mm.unlink();
+			} else {
+				wam.moveTo(SYNC_POS);
+
+				printf("Press [Enter] to link with the other Proficio.");
+				waitForEnter();
+				mm.tryLink();
+				wam.trackReferenceSignal(mm.output);
+
+				btsleep(0.1);  // wait an execution cycle or two
+				if (mm.isLinked()) {
+					printf("Linked with remote WAM.\n");
+				} else {
+					printf("WARNING: Linking was unsuccessful.\n");
+				}
+			}
+
+			break;
+
+		case 't':
+			size_t jointIndex;
+			{
+				size_t jointNumber;
+				std::cout << "\tJoint: ";
+				std::cin >> jointNumber;
+				jointIndex = jointNumber - 1;
+
+				if (jointIndex >= DOF) {
+					std::cout << "\tBad joint number: " << jointNumber;
+					break;
+				}
+			}
+
+			char gainId;
+			std::cout << "\tGain identifier (p, i, or d): ";
+			std::cin >> line;
+			gainId = line[0];
+
+			std::cout << "\tCurrent value: ";
+			switch (gainId) {
+			case 'p':
+				gainTmp = wam.jpController.getKp();
+				break;
+			case 'i':
+				gainTmp = wam.jpController.getKi();
+				break;
+			case 'd':
+				gainTmp = wam.jpController.getKd();
+				break;
+
+			default:
+				std::cout << "\tBad gain identifier.";
+			}
+			std::cout << gainTmp[jointIndex] << std::endl;
```

Comment:
This is less important, but better to skip this part if the user entered a bad gain identifier.